primitives/trie/src/trie_codec.rs

```diff
@@ -0,0 +1,184 @@
+/// an iterator of compact encoded nodes.
+///
+/// Child trie are decoded in order of child trie root present
+/// in the top trie.
+pub fn decode_compact<'a, L, DB, I>(
+	db: &mut DB,
+	encoded: I,
+	expected_root: Option<&TrieHash<L>>,
+) -> Result<TrieHash<L>, VerifyError<L>>
+	where
+		L: TrieConfiguration,
+		DB: HashDBT<L::Hash, trie_db::DBValue> + hash_db::HashDBRef<L::Hash, trie_db::DBValue>,
+		I: IntoIterator<Item = &'a [u8]>,
+{
+	let mut nodes_iter = encoded.into_iter();
+	let (top_root, _nb_used) = trie_db::decode_compact_from_iter::<L, _, _, _>(
+		db,
+		&mut nodes_iter,
+	).map_err(verify_error::<L>)?;
+
+	if let Some(expected_root) = expected_root {
+		if expected_root != &top_root {
+			return Err(VerifyError::<L>::RootMismatch(expected_root.clone()));
+		}
+	}
+
+	let mut child_tries = Vec::new();
+	{
+		// fetch child trie roots
+		let trie = crate::TrieDB::<L>::new(db, &top_root).map_err(verify_error::<L>)?;
+
+		use trie_db::Trie;
+		let mut iter = trie.iter().map_err(verify_error::<L>)?;
+
+		let childtrie_roots = sp_core::storage::well_known_keys::DEFAULT_CHILD_STORAGE_KEY_PREFIX;
+		if iter.seek(childtrie_roots).is_ok() {
+			loop {
+				match iter.next() {
+					Some(Ok((key, value))) if key.starts_with(childtrie_roots) => {
+						// we expect all default child trie root to be correctly encoded.
+						// see other child trie functions.
+						let mut root = TrieHash::<L>::default();
+						// still in a proof so prevent panic
+						if root.as_mut().len() != value.as_slice().len() {
+							return Err(VerifyError::<L>::RootMismatch(Default::default()));
+						}
+						root.as_mut().copy_from_slice(value.as_ref());
+						child_tries.push(root);
+					},
+					// allow incomplete database error: we only
+					// require access to data in the proof.
+					Some(Err(error)) => match *error {
+						trie_db::TrieError::IncompleteDatabase(..) => (),
+						e => return Err(VerifyError::<L>::DecodeError(Box::new(e))),
+					},
+					_ => break,
+				}
+			}
+		}
+	}
+
+	if !HashDBT::<L::Hash, _>::contains(db, &top_root, EMPTY_PREFIX) {
+		return Err(VerifyError::<L>::IncompleteProof);
+	}
+
+	let mut previous_extracted_child_trie = None;
+	for child_root in child_tries.into_iter() {
+		if previous_extracted_child_trie == None {
+			let (top_root, _) = trie_db::decode_compact_from_iter::<L, _, _, _>(
+				db,
+				&mut nodes_iter,
+			).map_err(verify_error::<L>)?;
+			previous_extracted_child_trie = Some(top_root);
+		}
+
+		// we allow skipping child root by only
+		// decoding next on match.
+		if Some(child_root) == previous_extracted_child_trie {
+			previous_extracted_child_trie = None;
+		}
+	}
+	if let Some(child_root) = previous_extracted_child_trie {
+		return Err(VerifyError::<L>::RootMismatch(child_root));
+	}
+
+	if nodes_iter.next().is_some() {
+		return Err(VerifyError::<L>::ExtraneousNode);
+	}
+
+	Ok(top_root)
+}
+
+pub fn encode_compact<'a, L>(
```

Comment:
Missing docs and also no tests.

Reply:
There is a slight test modification in state_machine lib.rs that test that. But could have one here too (in state machine we got already some test case with child trie so it is pretty convenient (the logic added in this pr is only the addition of child trie in the compact proof).

Reply:
You mean, there is testing in state-machine of this function, but I want one in the trie crate also?

Reply:
Or you want to have a separate testing function (I remember you don't like much test doing multiple things)?
Or you want to test the actual encoding (against a hard coded proof)?
@bkchr

Reply:
I mean no tests here in this crate. Maybe at least some encode/decode check